src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage.Accessor.cs

```diff
@@ -0,0 +1,317 @@
+                // database at all.
+                rowId = dataId;
+                return true;
+            }
+
+            protected virtual bool TryGetRowId(SqlConnection connection, Database database, TDatabaseId dataId, out long rowId)
+                => TryGetRowIdWorker(connection, database, dataId, out rowId);
+
+            private bool TryGetRowIdWorker(SqlConnection connection, Database database, TDatabaseId dataId, out long rowId)
+            {
+                // See https://sqlite.org/autoinc.html
+                // > In SQLite, table rows normally have a 64-bit signed integer ROWID which is 
+                // unique among all rows in the same table. (WITHOUT ROWID tables are the exception.)
+                // 
+                // You can access the ROWID of an SQLite table using one of the special column names 
+                // ROWID, _ROWID_, or OID. Except if you declare an ordinary table column to use one 
+                // of those special names, then the use of that name will refer to the declared column
+                // not to the internal ROWID.
+                using var resettableStatement = connection.GetResettableStatement(database == Database.WriteCache
+                    ? _select_rowid_from_writecache_table_where_0
+                    : _select_rowid_from_main_table_where_0);
+
+                var statement = resettableStatement.Statement;
+
+                BindFirstParameter(statement, dataId);
+
+                var stepResult = statement.Step();
+                if (stepResult == Result.ROW)
+                {
+                    rowId = statement.GetInt64At(columnIndex: 0);
+                    return true;
+                }
+
+                rowId = -1;
+                return false;
+            }
+
+            private void InsertOrReplaceBlobIntoWriteCache(
+                SqlConnection connection, TDatabaseId dataId,
+                byte[] checksumBytes, int checksumLength,
+                byte[] dataBytes, int dataLength)
+            {
+                using (var resettableStatement = connection.GetResettableStatement(_insert_or_replace_into_writecache_table_values_0_1_2))
+                {
+                    var statement = resettableStatement.Statement;
+
+                    // Binding indices are 1 based.
+                    BindFirstParameter(statement, dataId);
+                    statement.BindBlobParameter(parameterIndex: 2, value: checksumBytes, length: checksumLength);
+                    statement.BindBlobParameter(parameterIndex: 3, value: dataBytes, length: dataLength);
+
+                    statement.Step();
+                }
+
+                // Let the storage system know it should flush this information
+                // to disk in the future.
+                Storage.EnqueueFlushTask();
+            }
+
+            public void FlushInMemoryDataToDisk(SqlConnection connection)
```

Comment:
Looks like other methods were following a _MustRunInTransaction convention?

Reply:
Indeed.  Renamed for clarity.